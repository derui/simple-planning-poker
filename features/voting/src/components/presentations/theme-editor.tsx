import { Variant } from "@spp/shared-color-variant";
import { Icon, Icons } from "@spp/ui-icon";
import { clsx } from "clsx";
import { FormEvent, useEffect, useRef, useState } from "react";

export interface Props {
  /**
   * submit theme after editing finished
   */
  onSubmit?: (theme: string) => void;

  /**
   * Initial value of theme
   */
  theme: string;
}

const styles = {
  root: clsx("flex", "px-4", "py-1", "rounded-full", "border", "border-teal-600", "h-12"),
  contentContainer: clsx("flex-auto", "flex", "flex-row", "items-center", "gap-4"),
  theme: clsx("flex-auto", "text-teal-700", "font-bold", "text-lg"),
  themePlaceholder: clsx("flex-auto", "text-gray-700", "font-bold", "text-lg"),
  edit: clsx(
    "border",
    "border-transparent",
    "hover:border-orange-600",
    "hover:bg-orange-100",
    "transition",
    "p-1",
    "rounded-full"
  ),
  editor: {
    root: clsx("flex", "flex-row", "flex-auto", "items-center", "gap-2"),
    input: clsx("flex-auto", "w-full", "p-2", "outline-none", "rounded", "transition-colors"),
    submit: (editing: boolean) =>
      clsx(
        "flex-none",
        "border",
        "border-transparent",
        "hover:border-emerald-600",
        "hover:bg-emerald-100",
        "transition",
        "p-1",
        "rounded-full",
        {
          hidden: !editing,
        }
      ),
    cancel: (editing: boolean) =>
      clsx(
        "flex-none",
        "border",
        "border-transparent",
        "hover:border-gray-600",
        "hover:bg-gray-100",
        "transition",
        "p-1",
        "rounded-full",
        {
          hidden: !editing,
        }
      ),
  },
} as const;

const ThemeEditorInternal = function ThemeEditorInternal({
  editing,
  theme,
  onSubmit,
  onCancel,
}: Props & { editing: boolean; onCancel: () => void }) {
  const [state, setState] = useState(theme);
  const ref = useRef<HTMLInputElement>(null);

  useEffect(() => {
    if (ref.current) {
      ref.current.focus();
    }
  }, [ref.current]);

  const handleChange = (input: FormEvent<HTMLInputElement>) => {
    setState(input.currentTarget.value);
  };

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();

    if (editing) {
      onSubmit?.(state);
    }
  };

  return (
    <form className={styles.editor.root} onSubmit={handleSubmit}>
      <input
        ref={ref}
        type="text"
        tabIndex={1}
        className={styles.editor.input}
        onInput={handleChange}
        defaultValue={state}
        placeholder="No theme"
        readOnly={!editing}
      />
      <button type="submit" className={styles.editor.submit(editing)} disabled={!editing} aria-label="submit">
        <Icon type={Icons.check} variant={Variant.emerald} />
      </button>
      <button
        type="button"
        aria-label="cancel"
        className={styles.editor.cancel(editing)}
        onClick={onCancel}
        disabled={!editing}
      >
        <Icon type={Icons.x} variant={Variant.gray} />
      </button>
    </form>
  );
};

export const ThemeEditor = function ThemeEditor({ theme, onSubmit }: Props) {
  const [editing, setEditing] = useState(false);

  const handleEditing = () => {
    setEditing(true);
  };

  const handleCancel = () => {
    setEditing(false);
  };

  const handleSubmit = (value: string) => {
    setEditing(false);
    onSubmit?.(value);
  };

  const editButton = !editing ? (
    <button type="button" onClick={handleEditing} className={styles.edit} aria-label="edit">
      <Icon type={Icons.pencil} variant={Variant.orange} />
    </button>
  ) : null;

  return (
    <div className={styles.root}>
      <ThemeEditorInternal editing={editing} theme={theme} onSubmit={handleSubmit} onCancel={handleCancel} />
      {editButton}
    </div>
  );
};
