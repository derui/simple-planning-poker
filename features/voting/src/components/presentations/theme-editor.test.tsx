import { test, afterEach, expect } from "vitest";
import { render, cleanup, screen } from "@testing-library/react";
import { ThemeEditor } from "./theme-editor.js";
import { userEvent } from "@testing-library/user-event";

afterEach(cleanup);

test("should be able to render", () => {
  const ret = render(<ThemeEditor theme="foo" />);

  expect(ret.container).toMatchSnapshot();
});

test("show editor after click button", async () => {
  render(<ThemeEditor theme="foo" />);

  await userEvent.click(screen.getByRole("button", { name: "edit" }));

  expect(screen.queryAllByRole("button")).toHaveLength(2);
});

test("edit and submit changed theme", async () => {
  expect.assertions(1);
  render(
    <ThemeEditor
      theme="foo"
      onSubmit={(input) => {
        expect(input).toEqual("foo changed");
      }}
    />
  );

  await userEvent.click(screen.getByRole("button", { name: "edit" }));
  await userEvent.type(screen.getByPlaceholderText(/theme/), " changed[Enter]");
});

test("cancel should return default", async () => {
  render(<ThemeEditor theme="foo" />);

  await userEvent.click(screen.getByRole("button", { name: "edit" }));
  await userEvent.click(screen.getByRole("button", { name: "cancel" }));

  expect(screen.queryByPlaceholderText<HTMLInputElement>(/No theme/)?.value).toBe("foo");
});

test("focus input first when switched to edit mode", async () => {
  render(<ThemeEditor theme="foo" />);

  await userEvent.click(screen.getByRole("button", { name: "edit" }));

  expect(screen.getByRole<HTMLInputElement>("textbox").matches(":focus")).toBeTruthy();
});

test("should be able to render without theme", () => {
  const ret = render(<ThemeEditor theme="" />);

  expect(ret.container).toMatchSnapshot();
});
