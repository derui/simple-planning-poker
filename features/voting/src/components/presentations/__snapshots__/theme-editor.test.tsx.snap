// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should be able to render 1`] = `
<div>
  <div
    class="flex px-4 py-1 rounded-full border border-teal-600 h-12"
  >
    <form
      class="flex flex-row flex-auto items-center gap-2"
    >
      <input
        class="flex-auto w-full p-2 outline-none rounded transition-colors"
        placeholder="No theme"
        readonly=""
        tabindex="1"
        type="text"
        value="foo"
      />
      <button
        aria-label="submit"
        class="flex-none border border-transparent hover:border-emerald-600 hover:bg-emerald-100 transition p-1 rounded-full hidden"
        disabled=""
        type="submit"
      >
        <span
          class="flex before:bg-emerald-700 before:inline-block before:flex-none before:[mask-image:url(/static/svg/tabler-icons/check.svg)] before:[mask-size:cover] before:[mask-position:center] before:[mask-repeat:no-repeat] before:w-6 before:h-6 w-6 h-6"
        />
      </button>
      <button
        aria-label="cancel"
        class="flex-none border border-transparent hover:border-gray-600 hover:bg-gray-100 transition p-1 rounded-full hidden"
        disabled=""
        type="button"
      >
        <span
          class="flex before:bg-gray-700 before:inline-block before:flex-none before:[mask-image:url(/static/svg/tabler-icons/x.svg)] before:[mask-size:cover] before:[mask-position:center] before:[mask-repeat:no-repeat] before:w-6 before:h-6 w-6 h-6"
        />
      </button>
    </form>
    <button
      aria-label="edit"
      class="border border-transparent hover:border-orange-600 hover:bg-orange-100 transition p-1 rounded-full"
      type="button"
    >
      <span
        class="flex before:bg-orange-700 before:inline-block before:flex-none before:[mask-image:url(/static/svg/tabler-icons/pencil.svg)] before:[mask-size:cover] before:[mask-position:center] before:[mask-repeat:no-repeat] before:w-6 before:h-6 w-6 h-6"
      />
    </button>
  </div>
</div>
`;

exports[`should be able to render without theme 1`] = `
<div>
  <div
    class="flex px-4 py-1 rounded-full border border-teal-600 h-12"
  >
    <form
      class="flex flex-row flex-auto items-center gap-2"
    >
      <input
        class="flex-auto w-full p-2 outline-none rounded transition-colors"
        placeholder="No theme"
        readonly=""
        tabindex="1"
        type="text"
        value=""
      />
      <button
        aria-label="submit"
        class="flex-none border border-transparent hover:border-emerald-600 hover:bg-emerald-100 transition p-1 rounded-full hidden"
        disabled=""
        type="submit"
      >
        <span
          class="flex before:bg-emerald-700 before:inline-block before:flex-none before:[mask-image:url(/static/svg/tabler-icons/check.svg)] before:[mask-size:cover] before:[mask-position:center] before:[mask-repeat:no-repeat] before:w-6 before:h-6 w-6 h-6"
        />
      </button>
      <button
        aria-label="cancel"
        class="flex-none border border-transparent hover:border-gray-600 hover:bg-gray-100 transition p-1 rounded-full hidden"
        disabled=""
        type="button"
      >
        <span
          class="flex before:bg-gray-700 before:inline-block before:flex-none before:[mask-image:url(/static/svg/tabler-icons/x.svg)] before:[mask-size:cover] before:[mask-position:center] before:[mask-repeat:no-repeat] before:w-6 before:h-6 w-6 h-6"
        />
      </button>
    </form>
    <button
      aria-label="edit"
      class="border border-transparent hover:border-orange-600 hover:bg-orange-100 transition p-1 rounded-full"
      type="button"
    >
      <span
        class="flex before:bg-orange-700 before:inline-block before:flex-none before:[mask-image:url(/static/svg/tabler-icons/pencil.svg)] before:[mask-size:cover] before:[mask-position:center] before:[mask-repeat:no-repeat] before:w-6 before:h-6 w-6 h-6"
      />
    </button>
  </div>
</div>
`;
